### PR TITLE
Remove mutable_borrow_reservation_conflict lint and allow the code pattern

### LINKED ISSUE #56254: [NLL] prohibit "two-phase borrows" with existing borrows?
@RalfJung raised this example in which the "two-phase" borrow of `x` is compatible with a pre-existing share:

```rust
fn two_phase_overlapping1() {
    let mut x = vec![];
    let p = &x;
    x.push(p.len());
}
```

This poses a problem for stacked borrows, as well as for the potential refactoring of moving stacked borrows into MIR lowering (https://github.com/rust-lang/rust/issues/53198) -- roughly for the same reason. It might be nice to change this, but -- if so -- we've got to move quick!

cc @arielb1 @pnkfelix

### LINKED ISSUE #59159: Tracking issue for `mutable_borrow_reservation_conflict` compatibility lint
This is the **summary issue** for the `mutable_borrow_reservation_conflict` future-compatibility warning and other related errors. The goal of this page is describe why this change was made and how you can fix code that is affected by it. It also provides a place to ask questions or register a complaint if you feel the change should not be made. For more information on the policy around future-compatibility warnings, see our [breaking change policy guidelines][RFC 1589].

[RFC 1589]: https://github.com/rust-lang/rfcs/blob/master/text/1589-rustc-bug-fix-procedure.md

#### What is this lint about

A two-phase borrow is a mutable-borrow that is initially *reserved* at one point in the program's control-flow, and then subsequently *activated* at a later point in the control-flow.

For example, given a vector `v`, `v.push(v.len())` first reserves a borrow of `v` when it evaluates the method receiver, but does not activate that borrow until later when transferring control to the `push` method itself, *after* `v.len()` has been evaluated.

This lint detects instances where the reservation itself conflicts with some pre-existing borrow. For example:

```rust
   let mut v = vec![0, 1, 2];
   let shared = &v;
             // ~~ 
             // a shared borrow of `v` starts here ...
   v.push(shared.len());
// ~      ~~~~~~
// |      ... and that shared borrow is used here...
// |
// ... but that comes after the reservation of the
//     mutable borrow of `v` here (the reservation
//     is subsequently activated when `push` is invoked)
```

The latter code is an example of code that was accepted when two-phased borrows (2PB) initially landed, as part of non-lexical lifetimes (NLL) deployment in the 2018 edition.

This lint detects such cases, and warns that this pattern may be rejected by a future version of the compiler.

This is much further discussion of this at the following places:
 * Issue #56254
 * RalfJung blog post: https://www.ralfj.de/blog/2018/11/16/stacked-borrows-implementation.html
(The history section below attempts to provide a summary of the events that led to this lint being developed and the current status of the lint.)

#### How to fix this warning/error

Revise the code so that the initial evaluation of the mutable borrow (the "reservation") always comes after all uses of shared borrows it conflicts with.

One example revision of the example above:

```rust
let mut v = vec![0, 1, 2];
let shared = &v;
let len = shared.len();
v.push(len);
```

Now, the last use of `shared` comes *before* the reservation in `v.push(len)`, and thus there is no conflict between the shared borrow and the mutable borrow reservation.

#### Historical background 

At the time NLL was stabilized, this borrowing pattern was not meant to be accepted, because it complicates the abstract model for borrows and thus poses problems for unsafe code authors and for future compiler optimizations. (How *much* complication it introduces is a matter of debate, which is in part why this restriction is being given treatment different from other future compatibility lints.)

In other words: at the time that NLL was stabilized, the compiler's acceptance of this borrowing pattern was categorized by the NLL team as a "known bug". The NLL team assumed that, as a bug fix, the compiler would be allowed to start rejecting the pattern in the future.

Whether a future version of the compiler rejects the code will depend on an investigation of potential abstract models of the language semantics; we will not convert the lint to a hard error without first performing an evaluation of such abstract models. 

##### Timeline:

 * 2017-03-01 2PB [blog post](http://smallcultfollowing.com/babysteps/blog/2017/03/01/nested-method-calls-via-two-phase-borrowing/)
 * 2017-06-09 2PB RFC posted: https://github.com/rust-lang/rfcs/pull/2025
 * 2017-08-26 2PB RFC accepted and merged https://github.com/rust-lang/rfcs/pull/2025#issuecomment-325153662
 * 2017-12-06 2PB alpha implementation PRs: #46537, #47489, #48197
 * 2018-02-22 2PB alpha implementation identified as overly general and broken; new beta implementation proposed #48431 
 * 2018-03-06 2PB beta implementation via PR #48770
 * 2018-08-07: @RalfJung proposes "stacked borrows" semantic model for Rust borrowing and aliasing of references: [blog post](https://www.ralfj.de/blog/2018/08/07/stacked-borrows.html) [internals discussion](https://internals.rust-lang.org/t/stacked-borrows-an-aliasing-model-for-rust/8153)
 * 2018-11-16 stacked borrows implemented [blog post](https://www.ralfj.de/blog/2018/11/16/stacked-borrows-implementation.html)
 * 2018-11-26 @RalfJung points out that the stacked borrows model does not allow for pre-existing shares to overlap with reservations for a two-phase borrow #56254 
 * 2018-11-28 The NLL team decides that it should not attempt to rush to make a backportable change implementing the pre-existing share restriction. https://github.com/rust-lang/rust/issues/56254#issuecomment-442829785
 * 2018-11-29 The T-lang design team discussed the matter; it did not reach consensus about whether the NLL team was correct in its assumption that it could land the restriction later (by classifying the change as a "soundness fix"). https://github.com/rust-lang/rust/issues/56254#issuecomment-442980457
 * For the next four months, debate continued off and on about whether this borrowing pattern should be rejected or not.
 * 2019-02-25 Pull request implementing the proposed restriction is posted: PR #58739 
 * 2019-03-06 results from crater run with restriction posted https://github.com/rust-lang/rust/pull/58739#issuecomment-470230949
 * 2019-03-07 subset of T-lang design team discusses and @pnkfelix proposed PR #58739 be merged. https://github.com/rust-lang/rust/pull/58739#issuecomment-470676858 
 * For next three weeks, various people both within and apart from T-lang debate whether the restriction can be landed, given that NLL with a less-restricted 2PB was stabilized. Much of that debate is encoded in comments on PR #58739, such as this summary comment from niko: https://github.com/rust-lang/rust/pull/58739#issuecomment-473078780
 * 2019-03-25: compromise reached within T-lang design team: https://github.com/rust-lang/rust/pull/58739#issuecomment-476387184

#### Current status

- [x] PR #58739 introduces the `mutable_borrow_reservation_conflict` lint as warn-by-default
- [ ] PR #76104 makes the `mutable_borrow_reservation_conflict` lint deny-by-default
- [ ] PR ? makes the `mutable_borrow_reservation_conflict` lint a hard error

### PR DESCRIPTION
**Pull request by jackh726**
This was the only breaking issue with the NLL stabilization PR. Lang team decided to go ahead and allow this.

r? @nikomatsakis 
Closes #59159
Closes #56254

### CODE REVIEW

**Comment by jackh726** [comment]
Not sure if this needs FCP or not. There's technically no new behavior.

**Comment by pnkfelix** [comment]
I *really* want to go through the exercise of verifying that the current version of stacked-borrows does indeed allow the coding pattern.

Its not that I don't believe @RalfJung when they told me that it's fine and stacked-borrows has to handle this, but ... its still better to double check that key assumption first...

**Comment by pnkfelix** [comment]
Okay well I at least started digging into it and got far enough in the playground to convince myself that there's nothing in today's miri that is going to "obviously" explode. :)

More simply put, miri successfully executed this:

```rust
    #![allow(mutable_borrow_reservation_conflict)]
    let mut v = vec![0, 1, 2];
    let shared = &v;
    v.push(shared.len());
    dbg!(v);
```

so I'm a little less worried than I was 17 minutes ago.

**Comment by joshtriplett** [comment]
Let's go ahead and confirm consensus on this:

@rfcbot merge

**Comment by joshtriplett** [comment]
Let's go ahead and confirm consensus on this (momentarily switching labels to T-lang for rfcbot):

@rfcbot merge

**Comment by joshtriplett** [comment]
@rfcbot f? @RalfJung

**Comment by joshtriplett** [comment]
@rfcbot concern confirmation-from-ralf

@RalfJung, can we get explicit confirmation in this issue that Stacked Borrows won't have an issue with this anymore?

**Comment by jackh726** [comment]
Concern didn't go through...maybe because it's now tagged `T-compiler` again?

**Comment by cramertj** [comment]
@RalfJung 
> So as long as we are okay with (in the worst case) not doing any alias-based optimizations on 2phase-borrows, we are fine.

Does this mean that we would potentially plan to actually detect places where two-phase is not necessary and optimize only those places, or is this a de-facto "we're not going to do alias-based optimizations"?

If the former, that sounds scary. If the latter, it seems odd to me that one might have to revert to using non-two-phase-borrows-style code in order to unlock optimizations, but that seems like probably a niche enough occurrence that it wouldn't be a huge deal in practice? I have a very hard time judging the impact of these kinds of decisions. Are you able to shed any light on the kinds of optimizations that could be missed in common two-phase-borrows usecases?

**Comment by bjorn3** [comment]
How easy would it be to support this pattern if we were to move borrowck to THIR after we switch to polonius? (If moving it to THIR is possible in the first place that is.)

**Comment by RalfJung** [comment]
> Does this mean that we would potentially plan to actually detect places where two-phase is not necessary and optimize only those places, or is this a de-facto "we're not going to do alias-based optimizations"?

Many (most?) borrows are not two-phase borrows (including all the `noalias` we currently emit, since they are at function boundaries); those are unaffected.

But I admit I am not entirely familiar with the details of when two-phase borrows vs regular mutable borrows are used inside a function. (Shared references are fine anyway.)

The optimizations are the usual ones for aliasing guarantees -- reordering reads and write around other code. The [Stacked Borrows paper](https://plv.mpi-sws.org/rustbelt/stacked-borrows/) contains some examples.

> How easy would it be to support this pattern if we were to move borrowck to THIR after we switch to polonius? (If moving it to THIR is possible in the first place that is.)

Is this directed at me? The questions are about the run-time aliasing model, so borrowck has no bearing on this.

**Comment by saethlin** [comment]
How do we teach this? The only thing that I can imagine is that we now need to teach that "`&mut` requires uniqueness, except if you write code in _waves hands_ this sort of pattern, in that case `&mut` isn't unique".

That does not sound good.

**Comment by rdrpenguin04** [comment]
That's basically where I'm at.

Hello, I'm writing the borrow checker for LCCC, and this behavior seems underspecified. If someone could define why this pattern works, then I'd be all for allowing it; as it stands, this is a nightmare to be compatible with.

I strongly recommend holding off on removing this lint until we have a document specifying exactly why this is valid.

**Comment by asquared31415** [comment]
I would like to point out that it is always sufficient as far as I am aware to transform the code that requires this:
```rust
    let mut v = vec![0, 1, 2];
    let shared = &v;
    v.push(shared.len());
    dbg!(v);
```
into code that instead uses a temporary binding for the inner expression:
```rust
    let mut v = vec![0, 1, 2];
    let shared = &v;
    let inner = shared.len();
    v.push(inner);
    dbg!(v);
```
and there is precedent for requiring users to rearrange code to satisfy the borrow checker or other things, such as reordering unique borrows to not overlap or creating a binding to hold a temporary.

**Comment by rdrpenguin04** [comment]
To be fair, there is precedent for requiring reordering, but that's not something we should aspire to do. Once I learned the rules for activating borrows, it made sense; the only question now is teaching it, which should be doable.

**Comment by saethlin** [comment]
> So as long as we are okay with (in the worst case) not doing any alias-based optimizations on 2phase-borrows, we are fine.

@RalfJung does this mean it is incorrect to emit `noalias` for a `&mut` which is actually a 2-phase-borrow?

**Comment by nikomatsakis** [comment]
@rfcbot fcp reviewed

**Comment by RalfJung** [comment]
> does this mean it is incorrect to emit noalias for a &mut which is actually a 2-phase-borrow?

Yes, but we don't currently do that. Function arguments are always proper mutable borrows, on the callee side.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
r=me if the lint is deprecated "properly"

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 1]
```diff
@@ -3179,7 +3145,6 @@ declare_lint_pass! {
         META_VARIABLE_MISUSE,
         DEPRECATED_IN_FUTURE,
         AMBIGUOUS_ASSOCIATED_ITEMS,
-        MUTABLE_BORROW_RESERVATION_CONFLICT,
```
[the remark this answers is not in the document]
Comment (jackh726) [reply compiler/rustc_lint_defs/src/builtin.rs:3182]: It gets registered as removed here: https://github.com/rust-lang/rust/pull/96268/files#diff-cfb1d20a5949dd50b9769edb62eaf10a0de66734637fe71d35bdcf8fb06d767bR494

**Comment by jackh726** [comment]
@bors r=nikomatsakis rollup=iffy

Note to those that might try to add this to a rollup: tests *might* have changed that require a re-bless, but I'm not sure.

**Pull request merged**